File: mobile/lib/core/eligibility.dart
class EligibilityDecision {
  final String result; // eligible | temporarily_deferred | permanently_deferred
  final String? deferralReason;
  final DateTime? deferralUntil;

  const EligibilityDecision({
    required this.result,
    this.deferralReason,
    this.deferralUntil,
  });
}

class ScreeningInput {
  final int ageYears;
  final String sex;
  final double weightKg;
  final double hemoglobinGDl;
  final int bpSystolic;
  final int bpDiastolic;
  final int pulseBpm;
  final double temperatureCelsius;
  final bool hadRecentIllness;
  final bool hadRecentSurgery;
  final bool isPregnant;
  final bool hadTattooLast6m;
  final bool hadSti;
  final bool isOnMedication;
  final int? daysSinceLastDonation;

  const ScreeningInput({
    required this.ageYears,
    required this.sex,
    required this.weightKg,
    required this.hemoglobinGDl,
    required this.bpSystolic,
    required this.bpDiastolic,
    required this.pulseBpm,
    required this.temperatureCelsius,
    this.hadRecentIllness = false,
    this.hadRecentSurgery = false,
    this.isPregnant = false,
    this.hadTattooLast6m = false,
    this.hadSti = false,
    this.isOnMedication = false,
    this.daysSinceLastDonation,
  });
}

EligibilityDecision assessEligibility(ScreeningInput inp, {DateTime? today}) {
  final now = today ?? DateTime.now();

  EligibilityDecision temp(String reason, int days) => EligibilityDecision(
        result: 'temporarily_deferred',
        deferralReason: reason,
        deferralUntil: now.add(Duration(days: days)),
      );

  EligibilityDecision perm(String reason) => EligibilityDecision(
        result: 'permanently_deferred',
        deferralReason: reason,
      );

  if (inp.ageYears < 18) return perm('Age ${inp.ageYears} is below minimum of 18');
  if (inp.ageYears > 65) return perm('Age ${inp.ageYears} exceeds maximum of 65');
  if (inp.weightKg < 45) return temp('Weight ${inp.weightKg.toStringAsFixed(1)} kg below minimum 45 kg', 0);
  if (inp.hemoglobinGDl < 12.5) {
    return temp('Haemoglobin ${inp.hemoglobinGDl.toStringAsFixed(1)} g/dL below 12.5', 90);
  }
  if (inp.bpSystolic > 160 || inp.bpSystolic < 90) {
    return temp('Systolic BP ${inp.bpSystolic} out of range (90–160)', 7);
  }
  if (inp.bpDiastolic > 100 || inp.bpDiastolic < 50) {
    return temp('Diastolic BP ${inp.bpDiastolic} out of range (50–100)', 7);
  }
  if (inp.pulseBpm < 50 || inp.pulseBpm > 100) {
    return temp('Pulse ${inp.pulseBpm} out of range (50–100)', 7);
  }
  if (inp.temperatureCelsius > 37.5) {
    return temp('Temperature ${inp.temperatureCelsius.toStringAsFixed(1)} °C above 37.5', 14);
  }
  if (inp.isPregnant) return temp('Pregnancy or recent delivery/breastfeeding', 365);
  if (inp.hadRecentIllness) return temp('Recent illness in last 2 weeks', 14);
  if (inp.hadRecentSurgery) return temp('Surgery or dental procedure in last 6 months', 180);
  if (inp.hadTattooLast6m) return temp('Tattoo or body piercing in last 6 months', 180);
  if (inp.hadSti) return perm('History of sexually transmitted infection (STI)');
  if (inp.daysSinceLastDonation != null && inp.daysSinceLastDonation! < 90) {
    final left = 90 - inp.daysSinceLastDonation!;
    return temp(
      'Last donation was ${inp.daysSinceLastDonation} days ago (minimum 90 days required)',
      left,
    );
  }
  return const EligibilityDecision(result: 'eligible');
}

bool validateBarcode(String barcode) {
  if (barcode.length != 15) return false;
  const alphabet = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
  int charValue(String c) => alphabet.indexOf(c);
  final payload = barcode.substring(0, 14);
  final check = barcode.substring(14);
  var total = 0;
  final chars = payload.split('').reversed.toList();
  for (var i = 0; i < chars.length; i++) {
    var v = charValue(chars[i]);
    if (v < 0) return false;
    if (i % 2 == 0) {
      v *= 2;
      if (v > 35) v -= 35;
    }
    total += v;
  }
  final expected = alphabet[(36 - (total % 36)) % 36];
  return expected == check;
}

File: backend/app/core/eligibility.py
"""
Donor eligibility rules engine.

Rules are applied in priority order — first disqualifying rule wins.
Returns (EligibilityResult, deferral_reason, deferral_until).
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, timedelta

from app.models.enums import EligibilityResultEnum


@dataclass
class EligibilityDecision:
    result: EligibilityResultEnum
    deferral_reason: str | None = None
    deferral_until: date | None = None


@dataclass
class ScreeningInput:
    # Demographics
    age_years: int
    sex: str  # M / F / O
    weight_kg: float
    # Vitals
    hemoglobin_g_dl: float
    bp_systolic: int
    bp_diastolic: int
    pulse_bpm: int
    temperature_celsius: float
    # Questionnaire answers
    had_recent_illness: bool = False          # illness in last 2 weeks
    had_recent_surgery: bool = False          # surgery in last 6 months
    is_pregnant: bool = False                 # pregnancy / recent delivery
    had_tattoo_last_6m: bool = False          # tattoo or piercing in last 6 months
    had_sti: bool = False                     # STI history
    is_on_medication: bool = False            # on any long-term medication
    # Donation history (supplied by service layer)
    days_since_last_donation: int | None = None


_MIN_AGE = 18
_MAX_AGE = 65
_MIN_WEIGHT_KG = 45.0
_MIN_HB_GDL = 12.5          # applies to all; India NBTC standard
_MIN_INTERVAL_DAYS = 90     # 3 months between donations
_MAX_BP_SYSTOLIC = 160
_MIN_BP_SYSTOLIC = 90
_MAX_BP_DIASTOLIC = 100
_MIN_BP_DIASTOLIC = 50
_MIN_PULSE = 50
_MAX_PULSE = 100
_MAX_TEMP = 37.5


def _temp_defer(reason: str, days: int, today: date | None = None) -> EligibilityDecision:
    today = today or date.today()
    return EligibilityDecision(
        result=EligibilityResultEnum.TEMPORARILY_DEFERRED,
        deferral_reason=reason,
        deferral_until=today + timedelta(days=days),
    )


def _perm_defer(reason: str) -> EligibilityDecision:
    return EligibilityDecision(
        result=EligibilityResultEnum.PERMANENTLY_DEFERRED,
        deferral_reason=reason,
    )


def assess_eligibility(inp: ScreeningInput, today: date | None = None) -> EligibilityDecision:
    today = today or date.today()

    if inp.age_years < _MIN_AGE:
        return _perm_defer(f"Age {inp.age_years} is below minimum of {_MIN_AGE}")
    if inp.age_years > _MAX_AGE:
        return _perm_defer(f"Age {inp.age_years} exceeds maximum of {_MAX_AGE}")

    if inp.weight_kg < _MIN_WEIGHT_KG:
        return _temp_defer(f"Weight {inp.weight_kg:.1f} kg below minimum 45 kg", 0)

    if inp.hemoglobin_g_dl < _MIN_HB_GDL:
        return _temp_defer(f"Haemoglobin {inp.hemoglobin_g_dl:.1f} g/dL below 12.5", 90)

    if inp.bp_systolic > _MAX_BP_SYSTOLIC or inp.bp_systolic < _MIN_BP_SYSTOLIC:
        return _temp_defer(f"Systolic BP {inp.bp_systolic} out of range (90–160)", 7)
    if inp.bp_diastolic > _MAX_BP_DIASTOLIC or inp.bp_diastolic < _MIN_BP_DIASTOLIC:
        return _temp_defer(f"Diastolic BP {inp.bp_diastolic} out of range (50–100)", 7)

    if not (_MIN_PULSE <= inp.pulse_bpm <= _MAX_PULSE):
        return _temp_defer(f"Pulse {inp.pulse_bpm} out of range (50–100)", 7)

    if inp.temperature_celsius > _MAX_TEMP:
        return _temp_defer(f"Temperature {inp.temperature_celsius:.1f} °C above 37.5", 14)

    if inp.is_pregnant:
        return _temp_defer("Pregnancy or recent delivery/breastfeeding", 365)

    if inp.had_recent_illness:
        return _temp_defer("Recent illness in last 2 weeks", 14)

    if inp.had_recent_surgery:
        return _temp_defer("Surgery or dental procedure in last 6 months", 180)

    if inp.had_tattoo_last_6m:
        return _temp_defer("Tattoo or body piercing in last 6 months", 180)

    if inp.had_sti:
        return _perm_defer("History of sexually transmitted infection (STI)")

    if inp.days_since_last_donation is not None and inp.days_since_last_donation < _MIN_INTERVAL_DAYS:
        days_left = _MIN_INTERVAL_DAYS - inp.days_since_last_donation
        return _temp_defer(
            f"Last donation was {inp.days_since_last_donation} days ago (minimum 90 days required)",
            days_left,
            today,
        )

    return EligibilityDecision(result=EligibilityResultEnum.ELIGIBLE)
